### PR TITLE
Bugfix/2752 bulk editing removes categories

### LINKED ISSUE #2752: Bulk Editing of Release Dates of Programming Exercises Removes Categories
### Describe the bug
When using the "bulk editing" feature for programming exercises, which can be used to change release dates and deadlines, all categories of the selected programming exercises are lost. 

#### To Reproduce
1. In the instructor overview of all the exercises of one course, select one or several programming exercises via the check mark
2. Click on "Edit selected"
3. Change the dates and click on "Save"
4. If you go to the views of the individual exercises that you selected previously, you will see that the categories are now reset (i.e. there are none)

#### Expected behavior
The categories should not be affected by using the bulk edit feature.

### PR DESCRIPTION
**Pull request by daniels98it**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest2.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context

Fixes #2752 where categories of Programming Exercises are removed when 'bulk updated'.

In addition, the optional 'Notification Text' has been added to the 'Edit selected' modal, which was missing before.

### Description

A new Endpoint **PUT /programming-exercises/timeline** has been implemented to only update the timeline of a Programming Exercise. This fixes the before mentioned issue and improves the performance of the bulk updates.

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Select one or more Programming Exercises containing at least one 'Category'
4. Click on 'Edit Selected'
5. Choose a Release and Due date
6. Click 'Save'
7. The selected Exercises should now have the new release and due dates and their 'categories' should have remain untouched.

### Test Coverage

ProgrammingExerciseResourc.javae 82%
ProgrammingExerciseService.java 92%

programming-exercise-edit-selected.component.ts 87.27%
programming-exercise.service.ts 65.15%

### Screenshots

The *Edit Selected* modal with optional 'notification text field'
Screenshot from 2021-01-28 17-16-51

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java [hunk 1]
```diff
@@ -695,6 +695,42 @@ public void combineAllCommitsOfRepositoryIntoOne(VcsRepositoryUrl repoUrl) throw
...
+            programmingExercise.get().setDueDate(updatedProgrammingExercise.getDueDate());
+            programmingExercise.get().setBuildAndTestStudentSubmissionsAfterDueDate(updatedProgrammingExercise.getBuildAndTestStudentSubmissionsAfterDueDate());
+            programmingExercise.get().setAssessmentDueDate(updatedProgrammingExercise.getAssessmentDueDate());
+        }
+        else {
+            throw new EntityNotFoundException("Programming exercise not found with id: " + updatedProgrammingExercise.getId());
+        }
+
+        User user = userService.getUserWithGroupsAndAuthorities();
+        Course course = updatedProgrammingExercise.getCourseViaExerciseGroupOrCourseMember();
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java:721]: I think you should check this earlier

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 2]
```diff
@@ -688,6 +688,34 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
...
+    @FeatureToggle(Feature.PROGRAMMING_EXERCISES)
+    public ResponseEntity<ProgrammingExercise> updateProgrammingExerciseTimeline(@RequestBody ProgrammingExercise updatedProgrammingExercise,
+            @RequestParam(value = "notificationText", required = false) String notificationText) {
+        log.debug("REST request to update the timeline of ProgrammingExercise : {}", updatedProgrammingExercise);
+
+        try {
+            updatedProgrammingExercise = programmingExerciseService.updateTimeline(updatedProgrammingExercise, notificationText);
+        }
+        catch (IllegalAccessException ex) {
+            return forbidden();
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:709]: No need to catch those

**Review by krusche — CHANGES_REQUESTED** [review]

**Review by FlorianStang — APPROVED** [review]
Tested on TS3, everything worked as intended.

**Review by JannisBittlmayer — APPROVED** [review]
Tested on TS3, works as intended. Manual review and later steps are not set for the exercises when you define them but I assume that's intended.

**Review by airelawaleria — APPROVED** [review]

**Review by JohannesWeiss — DISMISSED** [review]
Works as expected on ts1

**Review by simonzilker — DISMISSED** [review]

**Review by philgabr — APPROVED** [review]
Works as intended :+1:

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java (hunk 1 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java]: Moved to ProgrammingExerciseResource.java with 
35a4022

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java (hunk 2 above)
Comment (daniels98it) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:709]: Catches removed with 
35a4022

**Review by marvinchimeka — DISMISSED** [review]
works as intended.

**Review by EneaGore — DISMISSED** [review]

**Comment by krusche** [comment]
> Tested on TS3, works as intended. Manual review and later steps are not set for the exercises when you define them but I assume that's intended.

If you can edit those in the client, I would assume that they are also changed in the database

**Review by valeriebucher — DISMISSED** [review]
Works as described!

**Review by DenizShabani — DISMISSED** [review]

**Comment by JannisBittlmayer** [comment]
> 
> 
> > Tested on TS3, works as intended. Manual review and later steps are not set for the exercises when you define them but I assume that's intended.
> 
> If you can edit those in the client, I would assume that they are also changed in the database

I think I formulated that too complicated. In short: You can activate the Manual Review in a bulk edit (see screenshot), but this does not activate the Manual Review in the exercises.
image

**Comment by daniels98it** [comment]
> > > Tested on TS3, works as intended. Manual review and later steps are not set for the exercises when you define them but I assume that's intended.
> > 
> > 
> > If you can edit those in the client, I would assume that they are also changed in the database
> 
> I think I formulated that too complicated. In short: You can activate the Manual Review in a bulk edit (see screenshot), but this does not activate the Manual Review in the exercises.
> image

You were right :+1: 
I was missing the Assessment Type attribute, which indeed did not update correctly.
Thanks for pointing that out :rocket:

**Review by sebastiantpeter — APPROVED** [review]
Tested on ts3. Issue seems to be fixed.

**Review by tiffca — APPROVED** [review]
Tested successfully

**Review by FrankeLukas — APPROVED** [review]
Changes lgtm and works on TS1.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java [hunk 3]
```diff
@@ -695,6 +695,33 @@ public void combineAllCommitsOfRepositoryIntoOne(VcsRepositoryUrl repoUrl) throw
...
+     * Updates the timeline attributes of the given programming exercise
+     * @param updatedProgrammingExercise containing the changes that have to be saved
+     * @param notificationText optional text for a notification to all students about the update
+     * @return the updated ProgrammingExercise object.
+     */
+    public ProgrammingExercise updateTimeline(ProgrammingExercise updatedProgrammingExercise, @Nullable String notificationText) {
+
+        Optional<ProgrammingExercise> programmingExercise = programmingExerciseRepository.findById(updatedProgrammingExercise.getId());
+        if (programmingExercise.isPresent()) {
+            programmingExercise.get().setReleaseDate(updatedProgrammingExercise.getReleaseDate());
```
Comment (ivanchimeno) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java:708]: You could technically avoid the multiple `.get()` calls by creating a local variable :D

Comment (ivanchimeno) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java:708]: You can do something like:
```java
Optional<ProgrammingExercise> programmingExerciseOrEmpty = programmingExerciseRepository.findById(updatedProgrammingExercise.getId());

if(programmingExerciseOrEmpty.isEmpty()) {
   throw new EntityNotFoundException("Programming exercise not found with id: " + updatedProgrammingExercise.getId());
}

var programmingExercise = programmingExerciseOrEmpty.get();

// The rest goes here

```

**Review by ivanchimeno — CHANGES_REQUESTED** [review]
Looks good to me. I just added a small code improvement.

**Review by krusche — APPROVED** [review]

**Pull request merged**